### PR TITLE
Draft PR for changes related to PA RTF merging

### PR DESCRIPTION
**Pull request by atharvasharma61**
**Is your feature request related to a problem? Please provide an existing Issue # , or describe.**

**Describe the solution you are proposing**

**Describe alternatives you've considered**

**Additional context**

### Check List
- [ ] Backport Labels added.
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/performance-analyzer/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by atharvasharma61** [comment]
https://pastecode.io/s/9bksb3ak 
Heap metrics

**Comment by ansjcy** [comment]
Hi @atharvasharma61! This looks like an interesting change, do we have any rfc/doc to provide more context?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+    private Histogram DiskWaitTimeMetrics;
+    private Histogram DiskServiceRateMetrics;
+    private Histogram DiskUtilizationMetrics;
+
+    private MetricsRegistry metricsRegistry;
+    private static final Logger LOG = LogManager.getLogger(RTFDisksCollector.class);
+
+    public RTFDisksCollector() {
+        super(
+                MetricsConfiguration.CONFIG_MAP.get(DisksCollector.class).samplingInterval,
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:35]: Why are we piggybacking on DisksCollector.class setting?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                metricsRegistry.createHistogram(
+                        AllMetrics.HeapValue.Constants.COLLECTION_COUNT_VALUE,
+                        "GC Collection Event PA Metrics",
+                        "1");
+
+        GCCollectionTimeMetrics =
+                metricsRegistry.createHistogram(
+                        AllMetrics.HeapValue.Constants.COLLECTION_TIME_VALUE,
+                        "GC Collection Time PA Metrics",
+                        "1");
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:74]: Why the unit is being marked "1" everywhere?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+import org.opensearch.telemetry.metrics.MetricsRegistry;
+import org.opensearch.telemetry.metrics.tags.Tags;
+
+public class RTFHeapMetricsCollector extends PerformanceAnalyzerMetricsCollector {
+    private static final Logger LOG = LogManager.getLogger(RTFHeapMetricsCollector.class);
+    public static final int SAMPLING_TIME_INTERVAL =
+            MetricsConfiguration.CONFIG_MAP.get(HeapMetricsCollector.class).samplingInterval;
+
+    private int count;
+    private Histogram GCCollectionEventMetrics;
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:32]: Everywhere in the code, please follow the java naming conventions.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+                    memoryUsage.getUsed(),
+                    Tags.create().addTag(MemTypeAttributeKey, entry.getKey()));
+        }
+
+        if (count == 12) {
+            count = 0;
+            for (Map.Entry<String, Supplier<MemoryUsage>> entry :
+                    HeapMetrics.getMemoryUsageSuppliers().entrySet()) {
+                MemoryUsage memoryUsage = entry.getValue().get();
+                metricsRegistry.createGauge(
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:110]: Gauge, first initialise once and then it automatically exports the latest value minutely(i.e. our export frequency) so should be ok to write every value to the guage if not aggregation is being done here.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFNodeStatsAllShardsMetricsCollector.java [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+import org.opensearch.telemetry.metrics.MetricsRegistry;
+import org.opensearch.telemetry.metrics.tags.Tags;
+
+public class RTFNodeStatsAllShardsMetricsCollector extends PerformanceAnalyzerMetricsCollector {
+    public static final int SAMPLING_TIME_INTERVAL =
+            MetricsConfiguration.CONFIG_MAP.get(NodeStatsAllShardsMetricsCollector.class)
+                    .samplingInterval;
+    private static final Logger LOG =
+            LogManager.getLogger(RTFNodeStatsAllShardsMetricsCollector.class);
+    private HashMap<ShardId, IndexShard> currentShards;
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFNodeStatsAllShardsMetricsCollector.java:42]: Define the Map only, HashMap could be an implementation that you want to use

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFNodeStatsAllShardsMetricsCollector.java [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+                            AllMetrics.ShardStatsValue.CACHE_REQUEST_EVICTION.toString(),
+                            (shardStats) -> shardStats.getStats().getRequestCache().getEvictions());
+                    put(
+                            AllMetrics.ShardStatsValue.CACHE_REQUEST_SIZE.toString(),
+                            (shardStats) ->
+                                    shardStats.getStats().getRequestCache().getMemorySizeInBytes());
+                }
+            };
+
+    private static final ImmutableMap<String, ValueCalculator> valueCalculators =
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFNodeStatsAllShardsMetricsCollector.java:124]: We can use ImmmutableMap directly instead of first creating the Map and then clone.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFNodeStatsAllShardsMetricsCollector.java [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+            return;
+        }
+
+        metricsRegistry = OpenSearchResources.INSTANCE.getMetricsRegistry();
+        if (metricsRegistry == null) {
+            LOG.error("could not get the instance of MetricsRegistry class");
+            return;
+        }
+
+        LOG.info("Executing collect metrics for RTFNodeStatsAllShardsMetricsCollector");
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFNodeStatsAllShardsMetricsCollector.java:140]: Make it debug.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFNodeStatsAllShardsMetricsCollector.java [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+        initialiseMetrics();
+        populateCurrentShards();
+        populatePerShardStats(indicesService);
+
+        for (HashMap.Entry currentShard : currentPerShardStats.entrySet()) {
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFNodeStatsAllShardsMetricsCollector.java:145]: Map.Entry

File: src/test/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPluginTests.java [hunk 9]
```diff
@@ -43,6 +43,8 @@
 import org.opensearch.plugins.ActionPlugin.ActionHandler;
```
Comment (Gaganjuneja) [inline src/test/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPluginTests.java:43]: Add tests for your collectors as well.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 4 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:116]: I looked into it but gauge does not provide us any method to reuse an already created gauge object.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java (hunk 1 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:35]: To match the collector periodicity. If needed, we can create new items in CONFIG_MAP for new collectors.

File: src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java [hunk 10]
```diff
@@ -227,6 +201,29 @@ public PerformanceAnalyzerPlugin(final Settings settings, final java.nio.file.Pa
                 new AdmissionControlMetricsCollector());
         scheduledMetricCollectorsExecutor.addScheduledMetricCollector(
                 new ElectionTermCollector(performanceAnalyzerController, configOverridesWrapper));
+        scheduledMetricCollectorsExecutor.addScheduledMetricCollector(new GCInfoCollector());
+
+        // Adding RTF Collectors if flag is enabled in performance-analyzer.properties
+        if (PluginSettings.instance().isTelemetryCollectorsEnabled()) {
```
[the remark this answers is not in the document]
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java:231]: IF PA plugin is disabled, scheduler will not schedule any collector. I will add another flag to control RCA collectors similar to how we are doing for telemetry collectors.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 4 above)
Comment (Gaganjuneja) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:116]: What did you mean by already created gauge object?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java [hunk 11]
```diff
@@ -54,52 +55,57 @@ public void collectMetrics(long startTime) {
             return;
         }
 
-        initialiseMetrics();
+        initialiseMetricsIfNeeded();
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:58]: Can we initialise in the constructor itself?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java (hunk 1 above)
Comment (Gaganjuneja) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:35]: Let's create new items please.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 11 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:58]: We can't. Metrics registry object is not set by the time these collectors are initialized. Collectors are initialized as the part of PerformanceAnalyzerPlugin constructor.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 4 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:116]: I might have misunderstood. Are you suggesting to avoid calling metricsRegistry.createGauge() everytime?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 11 above)
Comment (Gaganjuneja) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:58]: Ok.

File: src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java [hunk 12]
```diff
@@ -189,27 +195,19 @@ public PerformanceAnalyzerPlugin(final Settings settings, final java.nio.file.Pa
         clusterSettingsManager.addSubscriberForIntSetting(
                 PerformanceAnalyzerClusterSettings.PA_NODE_STATS_SETTING, nodeStatsSettingHandler);
 
-        scheduledMetricCollectorsExecutor.addScheduledMetricCollector(
-                new ThreadPoolMetricsCollector());
         scheduledMetricCollectorsExecutor.addScheduledMetricCollector(
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java:198]: Can we please extract out these collectors to 2 different methods like collectors and RTF collectors? I will be easy to switch over then.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 4 above)
Comment (Gaganjuneja) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:116]: Both metricsRegistry.createGauge() and counter handling. If you add multiple values to the Gauge metric it will only export the latest value added that's what I guess we are trying to do while emitting the 12th datapoint?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+                    metricsRegistry.createHistogram(
+                            AllMetrics.HeapValue.Constants.COLLECTION_COUNT_VALUE,
+                            "GC Collection Event PA Metrics",
+                            "");
+
+            gcCollectionTimeMetrics =
+                    metricsRegistry.createHistogram(
+                            AllMetrics.HeapValue.Constants.COLLECTION_TIME_VALUE,
+                            "GC Collection Time PA Metrics",
+                            "");
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:76]: unit here could be "ms"

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+        LOG.debug("Executing collect metrics for RTFDisksCollector");
+
+        initialiseMetricsIfNeeded();
+        DiskMetricsGenerator diskMetricsGenerator = generator.getDiskMetricsGenerator();
+        diskMetricsGenerator.addSample();
+
+        recordMetrics(diskMetricsGenerator);
+    }
+
+    private void recordMetrics(DiskMetricsGenerator diskMetricsGenerator) {
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:65]: Please make sure the metrics and dimension names are in sync with the existing system.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 4 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:116]: Ack. Then I will remove the if block.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java (hunk 14 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:65]: Yes. Metrics and dimension names are unchanged.

File: src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java [hunk 15]
```diff
@@ -189,27 +195,61 @@ public PerformanceAnalyzerPlugin(final Settings settings, final java.nio.file.Pa
...
+
+        // Adding RCA Collectors if flag is enabled in performance-analyzer.properties
+        if (PluginSettings.instance().isRcaCollectorsEnabled()) {
+            LOG.info("Scheduling RCA Collectors");
+            scheduleRcaCollectors();
+        }
+
+        scheduledMetricCollectorsExecutor.start();
+
+        EventLog eventLog = new EventLog();
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java:212]: Can we do this also behind a flag?

File: src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java [hunk 16]
```diff
@@ -189,27 +195,61 @@ public PerformanceAnalyzerPlugin(final Settings settings, final java.nio.file.Pa
...
                 PerformanceAnalyzerClusterSettings.PA_NODE_STATS_SETTING, nodeStatsSettingHandler);
 
+        // Adding RTF Collectors if flag is enabled in performance-analyzer.properties
+        if (PluginSettings.instance().isTelemetryCollectorsEnabled()) {
+            LOG.info("Scheduling Telemetry Collectors");
+            scheduleTelemetryCollectors();
+        }
+
+        // Adding RCA Collectors if flag is enabled in performance-analyzer.properties
+        if (PluginSettings.instance().isRcaCollectorsEnabled()) {
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java:205]: What will happen to the delta metrics if 2 collectors are running parallel in both modes?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        for (String disk : diskMetricsGenerator.getAllDisks()) {
+            Tags diskNameTag = Tags.create().addTag("disk_name", disk);
+            double Disk_WaitTime = diskMetricsGenerator.getAwait(disk);
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:67]: Follow the java camelCase naming convention.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java [hunk 18]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        initialiseMetricsIfNeeded();
+        GCMetrics.runGCMetrics();
+        LOG.debug("Executing collect metrics for RTFHeapMetricsCollector");
+        recordMetrics();
+    }
+
+    private void initialiseMetricsIfNeeded() {
+        if (metricsInitialised == false) {
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:60]: I am wondering should we have a flag or not null check? Because if we may miss the setting flag to null then it will be an issue and hard to track.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+
+        gcCollectionTimeMetrics.record(GCMetrics.getTotFullGCCollectionTime(), totFullGCTag);
+
+        for (Map.Entry<String, Supplier<MemoryUsage>> entry :
+                HeapMetrics.getMemoryUsageSuppliers().entrySet()) {
+            MemoryUsage memoryUsage = entry.getValue().get();
+            heapUsedMetrics.record(
+                    memoryUsage.getUsed(),
+                    Tags.create().addTag(memTypeAttributeKey, entry.getKey()));
+            metricsRegistry.createGauge(
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:104]: Gauge should also be created once.

File: src/test/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPluginTests.java [hunk 20]
```diff
@@ -78,7 +83,19 @@ public void setup() {
...
+            metricsRegistryFactory.close();
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        clusterService =
+                new ClusterService(
+                        settings,
+                        clusterSettings,
+                        threadPool,
+                        new ClusterManagerMetrics(metricsRegistry));
```
Comment (Gaganjuneja) [inline src/test/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPluginTests.java:98]: Do we need this change now?

File: src/test/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollectorTests.java [hunk 21]
```diff
@@ -0,0 +1,71 @@
...
+        System.setProperty("performanceanalyzer.metrics.log.enabled", "False");
+        MetricsConfiguration.CONFIG_MAP.put(RTFDisksCollector.class, MetricsConfiguration.cdefault);
+
+        metricsRegistry = mock(MetricsRegistry.class);
+        diskWaitTimeHistogram = mock(Histogram.class);
+        diskServiceRateHistogram = mock(Histogram.class);
+        diskUtilizationHistogram = mock(Histogram.class);
+        OpenSearchResources.INSTANCE.setMetricsRegistry(metricsRegistry);
+
+        when(metricsRegistry.createHistogram(anyString(), anyString(), anyString()))
```
Comment (Gaganjuneja) [inline src/test/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollectorTests.java:42]: Can we make it more predictable and simply use thenReturn to validate for all type of histograms. Here `diskUtilizationHistogram` will only be returned always.

File: src/test/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollectorTests.java [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+
+    private static MetricsRegistry metricsRegistry;
+    private static Histogram gcCollectionEventHistogram;
+    private static Histogram gcCollectionTimeHistogram;
+    private static Histogram heapUsedHistogram;
+
+    @Before
+    public void init() {
+        initMocks(this);
+        System.setProperty("performanceanalyzer.metrics.log.enabled", "False");
```
Comment (Gaganjuneja) [inline src/test/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollectorTests.java:35]: Can we do this in one base test class?

File: src/test/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFThreadPoolMetricsCollectorTests.java [hunk 23]
```diff
@@ -0,0 +1,92 @@
...
+                                    AllMetrics.ThreadPoolValue.Constants.QUEUE_CAPACITY_VALUE)) {
+                                return threadPoolQueueCapacityHistogram;
+                            }
+                            return threadPoolTotalThreadsHistogram;
+                        });
+
+        rtfThreadPoolMetricsCollector = new RTFThreadPoolMetricsCollector();
+    }
+
+    @Test
```
Comment (Gaganjuneja) [inline src/test/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFThreadPoolMetricsCollectorTests.java:76]: Add tests for negative cases as well wherever possible.

**Review by Gaganjuneja — CHANGES_REQUESTED** [review]

File: src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java (hunk 16 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java:205]: previous state and current state are maintained inside the collectors and diff is calculated there only. This should not be a concern since we have different collectors for RTF vs non-RTF collectors.

File: src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java (hunk 15 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/PerformanceAnalyzerPlugin.java:196]: We should not do it since we are now dynamically controlling collectors. EventLog should be always initialised if we wish to control the changes dynamically, otherwise it will require a B/G.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 18 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:79]: Would like to discuss this one.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 19 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:123]: I tried creating the gauge object once and storing it but it does not provide any method that could allow us to re-use an existing object for measurements.

File: src/main/java/org/opensearch/performanceanalyzer/config/setting/PerformanceAnalyzerClusterSettings.java [hunk 24]
```diff
@@ -39,6 +39,13 @@ public enum PerformanceAnalyzerFeatureBits {
                     Setting.Property.NodeScope,
                     Setting.Property.Dynamic);
 
+    public static final Setting<Integer> PA_COLLECTORS_SETTING =
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/config/setting/PerformanceAnalyzerClusterSettings.java:42]: Please provide the java doc with different setting values possible

File: src/main/java/org/opensearch/performanceanalyzer/config/setting/handler/PerformanceAnalyzerCollectorsSettingHandler.java [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+ */
+
+package org.opensearch.performanceanalyzer.config.setting.handler;
+
+import org.opensearch.performanceanalyzer.config.PerformanceAnalyzerController;
+import org.opensearch.performanceanalyzer.config.setting.ClusterSettingListener;
+import org.opensearch.performanceanalyzer.config.setting.ClusterSettingsManager;
+import org.opensearch.performanceanalyzer.config.setting.PerformanceAnalyzerClusterSettings;
+
+public class PerformanceAnalyzerCollectorsSettingHandler
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/config/setting/handler/PerformanceAnalyzerCollectorsSettingHandler.java:13]: Pls provide java docs.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 19 above)
Comment (Gaganjuneja) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:120]: Gauge accepts one Supplier which will be executed once in a defined period which is default 1 min in OpenSearch core.

**Review by Gaganjuneja — CHANGES_REQUESTED** [review]

**Comment by Gaganjuneja** [comment]
Overall looks good, please address the open comments and build failures.

**Comment by Gaganjuneja** [comment]
I am hoping there are changes in `performance-analyzer-commons` package also in class `ScheduledMetricCollectorsExecutor`. Please attache that PR also here.

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 18 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:76]: Discussed offline. We should be good with this implementation for now. We will revisit if needed.

File: src/main/java/org/opensearch/performanceanalyzer/config/setting/handler/PerformanceAnalyzerCollectorsSettingHandler.java (hunk 25 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/config/setting/handler/PerformanceAnalyzerCollectorsSettingHandler.java:17]: Added

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 19 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:120]: But tags can only be added while creating gauge. That would be a problem in your approach.

**Comment by atharvasharma61** [comment]
One more thing to call-out: We need to enable delta temporality in Opensearch Otel. @Gaganjuneja

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java (hunk 19 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFHeapMetricsCollector.java:122]: Discussed offline to keep it as it is until further development

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+    private PerformanceAnalyzerController performanceAnalyzerController;
+    private ConfigOverridesWrapper configOverridesWrapper;
+
+    public RTFDisksCollector(
+            PerformanceAnalyzerController performanceAnalyzerController,
+            ConfigOverridesWrapper configOverridesWrapper) {
+        super(
+                MetricsConfiguration.CONFIG_MAP.get(RTFDisksCollector.class).samplingInterval,
+                "RTFDisksCollector",
+                StatMetrics.DISKS_COLLECTOR_EXECUTION_TIME,
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:44]: Why are we publishing the stats against the existing RCA collector metrics?

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java [hunk 27]
```diff
@@ -0,0 +1,113 @@
...
+            diskServiceRateMetrics.record(diskServiceRate, diskNameTag);
+        }
+    }
+
+    private void initialiseMetricsIfNeeded() {
+        if (metricsInitialised == false) {
+            diskWaitTimeMetrics =
+                    metricsRegistry.createHistogram(
+                            RTFMetrics.DiskValue.Constants.WAIT_VALUE,
+                            "DiskWaitTimeMetrics",
```
Comment (Gaganjuneja) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:98]: Please follow the otel convention for metric names as well.

Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:98]: That's metrics description. Metric name is the first parameter

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java (hunk 26 above)
Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:44]: This is for PMET. I can update it but it's not required since it is dependent on RCA from what I understand. We need to look into that statsCollector but I would still update this one for consistency.

**Review by Gaganjuneja — APPROVED** [review]

**Comment by Gaganjuneja** [comment]
Please take a look at build failures.

**Comment by atharvasharma61** [comment]
> Please take a look at build failures.

Dependent on PRs: https://github.com/opensearch-project/performance-analyzer/pull/657 and https://github.com/opensearch-project/performance-analyzer-commons/pull/74

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java [hunk 28]
```diff
@@ -0,0 +1,113 @@
...
+import org.opensearch.performanceanalyzer.commons.metrics_generator.DiskMetricsGenerator;
+import org.opensearch.performanceanalyzer.commons.metrics_generator.OSMetricsGenerator;
+import org.opensearch.performanceanalyzer.commons.stats.metrics.StatExceptionCode;
+import org.opensearch.performanceanalyzer.commons.stats.metrics.StatMetrics;
+import org.opensearch.performanceanalyzer.config.PerformanceAnalyzerController;
+import org.opensearch.telemetry.metrics.Histogram;
+import org.opensearch.telemetry.metrics.MetricsRegistry;
+import org.opensearch.telemetry.metrics.tags.Tags;
+
+public class RTFDisksCollector extends PerformanceAnalyzerMetricsCollector
```
Comment (nishchay21) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:26]: please add java doc

File: src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+
+        metricsRegistry = OpenSearchResources.INSTANCE.getMetricsRegistry();
+        if (metricsRegistry == null) {
+            LOG.error("could not get the instance of MetricsRegistry class");
+            return;
+        }
+
+        LOG.debug("Executing collect metrics for RTFDisksCollector");
+
+        initialiseMetricsIfNeeded();
```
Comment (nishchay21) [inline src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:73]: Do we need to call this again can we not use the constructor for this

Comment (atharvasharma61) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:73]: This cannot be added in constructor because metricsRegistry object is not available by that time. This is already covered in one of the resolved comments.

Comment (nishchay21) [reply src/main/java/org/opensearch/performanceanalyzer/collectors/telemetry/RTFDisksCollector.java:73]: For now should be ok lets see if we can improvise on this as P1

**Review by nishchay21 — DISMISSED** [review]

**Review by nishchay21 — APPROVED** [review]

**Review by Gaganjuneja — APPROVED** [review]
